mersenne-31/src/extension2.rs

```diff
@@ -0,0 +1,42 @@
+use core::ops::Mul;
+
+use p3_field::{define_ext, impl_ext_af, impl_ext_f, AbstractField, Field};
+
+use crate::Mersenne31;
+
+define_ext!(pub CM31([Mersenne31; 2]));
+
+impl<AF: AbstractField<F = Mersenne31>> Mul for CM31<AF> {
+    type Output = Self;
+    fn mul(self, rhs: Self) -> Self::Output {
+        let ([a, b], [c, d]) = (self.0, rhs.0);
+        Self([
+            a.clone() * c.clone() - b.clone() * d.clone(),
+            a.clone() * d.clone() + b.clone() * c.clone(),
+        ])
+    }
+}
+
+impl<AF: AbstractField<F = Mersenne31>> AbstractField for CM31<AF> {
+    impl_ext_af!();
+    type F = CM31<Mersenne31>;
+    fn generator() -> Self {
+        Self([Mersenne31::new(12), Mersenne31::new(1)].map(AF::from_f))
+    }
```

Comment:
Just leaving a note to myself to double check this in sage. (Will do it tomorrow)

Reply:
Confirmed with:
p = 2^31 - 1
F = GF(p)
CM31 = F.extension(x^2 + 1, 'x')
CM31.multiplicative_generator()
(The last function returns x + 12).